### ISSUE TITLE
Introduce a matrix_inverse_transpose for matrix3x4

### ISSUE THREAD

**Issue by nfrechette** [nfrechette/rtm#10]
The most common usage of the matrix inverse transpose for affine matrices is to properly deform normals. An optimized implementation should be added. It should return a matrix3x3 since the transpose returns that as well.

See here for details: https://paroj.github.io/gltut/Illumination/Tut09%20Normal%20Transformation.html

**Comment by Nehon** [comment]
Hello, maybe you should look into this
https://github.com/graphitemaster/normals_revisited
and this
https://www.shadertoy.com/view/3s33zj

TL;DR is: using the inverse transpose for normal deformation is wrong in many cases, especially when you have non uniform scales.
The proposed solution here is to use the adjoint matrix (usually known as the adjugate), it's right and faster to compute.

All this to say that, maybe you should look into adding an adjugate() function instead of a inverse_transpose, if the goal is to provide a fast method to compute a matrix to deform normals.

**Comment by nfrechette** [comment]
Very interesting, thanks! Adding an adjugate indeed seems to make sense but his derivation is a bit hard to follow. The main difference, aside from speed, seems to be when a single (or all 3) axis have negative scale. In such a setup, the determinant will be negative since winding is reversed which will flip the resulting normal. I'll have to play with it a bit. I've seen the transpose inverse used a lot but I've never seen the adjugate used.

I'll have to double check this to better understand what happens with the edge cases.

**Comment by nfrechette** [comment]
I looked into details and as @Nehon mentions, transforming a normal by the cofactor matrix is the correct way (as described in the normals_revisited link). I've added support for calculating matrix minors, cofactors, determinants, and the adjugate.

Note that the adjugate is the transpose of the cofactor matrix and that the shader toy example does not calculate the adjugate but the cofactor matrix. I've generated random matrices and used Wolfram Alpha to calculate reference values for the unit tests.

This will be merged in develop later today if everything goes well.
